src/main/java/org/testng/reporters/XMLSuiteResultWriter.java

```diff
@@ -284,23 +284,39 @@ private void addTestResultException(XMLStringBuffer xmlBuffer, ITestResult testR
       }
 
       String[] stackTraces = Utils.stackTrace(exception, false);
-      if ((config.getStackTraceOutputMethod() & XMLReporterConfig.STACKTRACE_SHORT) == XMLReporterConfig
-              .STACKTRACE_SHORT) {
-        xmlBuffer.push(XMLReporterConfig.TAG_SHORT_STACKTRACE);
-        xmlBuffer.addCDATA(stackTraces[0]);
-        xmlBuffer.pop();
-      }
-      if ((config.getStackTraceOutputMethod() & XMLReporterConfig.STACKTRACE_FULL) == XMLReporterConfig
-              .STACKTRACE_FULL) {
-        xmlBuffer.push(XMLReporterConfig.TAG_FULL_STACKTRACE);
-        xmlBuffer.addCDATA(stackTraces[1]);
-        xmlBuffer.pop();
+      int level = calculateStackTraceLevels(testResult);
+      switch (level) {
+        case XMLReporterConfig.STACKTRACE_SHORT:
+          xmlBuffer.push(XMLReporterConfig.TAG_SHORT_STACKTRACE);
+          xmlBuffer.addCDATA(stackTraces[0]);
```

Comment:
Isn't `stackTraces[0]` (and `[1]`) just one frame of the stack trace? I don't understand what makes them short or full...